patches/tModLoader/Terraria/ModLoader/NPCLoader.cs

```diff
@@ -35,12 +35,32 @@ public static class NPCLoader
 
 		private class HookList
 		{
-			public GlobalNPC[] arr = new GlobalNPC[0];
+			public int[] arr = new int[0];
 			public readonly MethodInfo method;
 
 			public HookList(MethodInfo method) {
 				this.method = method;
 			}
+
+			public IEnumerable<GlobalNPC> Enumerate(NPC npc) {
```

Comment:
I've long since banned anything except array foreach from tML in hooks. Because of allocations and the inability for jit to optimise it. 

Specifically, if we're at the point where we want to avoid empty function call overhead (the whole point of building hooklists) then we're on the same order of magnitude as non-optimised loops. Seriously, instruction pre-fetch is mandatory for modern day performance.

That said, C# _recently_ added support for structs implementing IEnumerator. There's a good chance that a simple enough one could be optimised by the JIT and it won't allocate, so, try it!

Reply:
Erm, if you're so willing, could you elaborate as to what this means? I'm just really curious. It doesn't really pertain to me otherwise, but what makes non-array foreach not JIT optimizable, why is that related to allocation, and what's instruction pre-fetch?